src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs

```diff
@@ -115,105 +113,147 @@ public async Task<Solution> CreateSolutionAsync(Checksum newSolutionChecksum, Ca
+                // desirable as it will likely be useful in future calls to still know about that project info without
+                // it being dropped and having to be resynced.
+                var isConeSync = newSolutionChecksums.ProjectConeId != null;
+                if (isConeSync)
                 {
-                    if (!oldMap.ContainsKey(projectId))
+                    foreach (var (oldChecksum, oldProjectId) in oldSolutionChecksums.Projects)
                     {
-                        var projectInfo = await _assetProvider.CreateProjectInfoAsync(projectId, newProjectChecksums.Checksum, cancellationToken).ConfigureAwait(false);
-                        solution = solution.AddProject(projectInfo);
+                        if (!newProjectIdToChecksum.ContainsKey(oldProjectId))
+                            oldProjectIdToChecksum.Remove(oldProjectId);
                     }
+
+                    // All the old projects must be in the new project set.  Though the reverse doesn't have to hold.
+                    // The new project set may contain additional projects to add.
+                    Contract.ThrowIfFalse(oldProjectIdToChecksum.Keys.All(newProjectIdToChecksum.Keys.Contains));
                 }
 
-                // remove all project references from projects that changed. this ensures exceptions will not occur for
-                // cyclic references during an incremental update.
-                foreach (var (projectId, newProjectChecksums) in newMap)
+                // remove projects that are the same on both sides.
+                foreach (var projectId in allProjectIds)
                 {
-                    if (!oldMap.TryGetValue(projectId, out var oldProjectChecksums))
+                    if (oldProjectIdToChecksum.TryGetValue(projectId, out var oldChecksum) &&
+                        newProjectIdToChecksum.TryGetValue(projectId, out var newChecksum) &&
+                        oldChecksum == newChecksum)
                     {
-                        continue;
-                    }
-
-                    if (oldProjectChecksums.ProjectReferences.Checksum != newProjectChecksums.ProjectReferences.Checksum)
-                    {
-                        solution = solution.WithProjectReferences(projectId, SpecializedCollections.EmptyEnumerable<ProjectReference>());
+                        oldProjectIdToChecksum.Remove(projectId);
+                        newProjectIdToChecksum.Remove(projectId);
                     }
                 }
 
-                // removed project
-                foreach (var (projectId, _) in oldMap)
+                using var _4 = PooledDictionary<ProjectId, ProjectStateChecksums>.GetInstance(out var oldProjectIdToStateChecksums);
+                using var _5 = PooledDictionary<ProjectId, ProjectStateChecksums>.GetInstance(out var newProjectIdToStateChecksums);
```

Comment:
first pass is just projectIds+checksums.  next pass is projectids+statechecksums.  Recall that statechecksums are the rich tree objects that then break down all the individual checksum pieces of a particular object (in this case a project).

So, once we see that a project is actually changed (it's raw checksum is different), we collect these rich objects to find out what was actually changed in order to sync that info and update it.